### ISSUE TITLE
Converting Money to JSON and casting it back results in a different value

### ISSUE THREAD

**Issue by jfpedroza** [kipcole9/money#145]
I deal with COP most of the time, and COP uses `,` for decimals and `.` for thousands.

Now, the issue is that when converting to JSON, the amount uses `.` for decimals.

For example:

```elixir
m = Money.new(:COP, "1000,25")
%{"currency" => currency, "amount" => amount} = Jason.encode!(m) |> Jason.decode!()
Money.new(currency, amount)
```

Results in `Money.new(:COP, "100025")`

That means the amount cannot be passed directly to `Money.new` or the map directly to a changeset.

What would be the best way to handle this? Besides creating a Decimal before passing it to `Money.new` or `Ecto.Changeset.cast`.

**Comment by kipcole9** [comment]
Thanks for the issue @jfpedroza, its a good discussion.  I do sanitize the strings in `Money.new/2` before passing them to `Decimal.new/1` as you've seen. And that sanitisation is not down in a locale-aware fashion.  Thats a bug and I appreciate you bringing it to my attention (its never come up before!). I will work on this but it will take until the weekend for me to get to it most likely.

I believe the JSON format is the correct one - its not meant to be human formatted, its a machine representation. I believe it should only use `.` as the decimal separator and have no grouping separator.

If you need to parse strings that contain currencies then you can parse the string since parsing is locale aware:
```elixir
# Parse a full money string
iex> Money.parse("COP 1000,25", locale: "es-CO")
Money.new(:COP, "1000.25")

# Parse just the money amount
iex> Cldr.Number.Parser.parse "1000,25", locale: "es-CO", number: :decimal
{:ok, #Decimal<1000.25>}
```

In summary:
1. I believe the JSON format is the correct one. Its intended for machine reading/writing, not for humans
2. `Money.new/2` should parse string amounts in a locale aware fashion. I will fix this, but it will take a few days
3. You can parse money strings, or number string, in a locale-aware fashion.

Thoughts?

**Comment by jfpedroza** [comment]
I agree that JSON stay as is as it is indeed intended for machines.

How do you recommend the case where we have the map received from JSON?

A case that happens often is passing the Money struct to an Oban job and that is stored as JSON. Then receiving the map in the worker and wanting to convert it back to a Money struct.

Should we do this?

```elixir
%{"currency" => currency, "amount" => amount} = map
Money.new(currency, Decimal.new(amount)
```

It's a bit more difficult when you have a nested structure and use `Money.Ecto.Composite.Type ` in a changeset though.

And with the fix, how would passing `1000.25` to `Money.new` behave? Same as `1000,25` or fail?

**Comment by kipcole9** [comment]
I'm a little bit confused now so excuse me if I've misunderstood.

* `assert Money.new(:COP, "100.25") |> Jason.encode!() |> Jason.decode!() |> Money.Ecto.Composite.Type.cast()) == Money.new(:COP, "100.25")` should always be true. Are you seeing a situation where that is not true? 

* Unfortunately it's not possible to `type` a JSON object so I can't tell if the map comes from converting a `Money.t` struct or is some arbitrary map. Therefore if you are saying you have a map (converted from JSON) like `%{currency: "COP", amount: "100,25"}` then you'll need to do `Money.new(map.currency, Cldr.Number.Parser.parse(map.amount, locale: "es-CO", number: :decimal))` or similar.

Did I miss something?

**Comment by jfpedroza** [comment]
1. It's true, just that passing "100.25" results in "10025".
2. The map will be `%{"currency" => "COP", "amount" => "100.25"}` (with `.`) because it comes from JSON.

**Comment by kipcole9** [comment]
> And with the fix, how would passing 1000.25 to Money.new behave? Same as 1000,25 or fail?

The fix isn't required, `Money.new/3` is already locale-aware. (my bad).  All parsing is done in a locale-aware fashion and the default locale is `Cldr.get_locale/0` (see the `:locale` option below):

## Money.new/3

### Arguments

* currency_code is an ISO4217 (https://en.wikipedia.org/wiki/ISO_4217)
    binary or atom currency code or an ISO 24165
    (https://www.iso.org/standard/80601.html) token identifier or shortname.

* amount is an integer, string or Decimal money amount.

* options is a keyword list of options.

### Options

* :locale is any known locale.  The locale is used to normalize any
    binary (String) amounts to a form that can be consumed by Decimal.new/1.
    This consists of removing any localised grouping characters and replacing
    the localised decimal separator with a ".". The default is
    Cldr.get_locale/0.

* :backend is any module() that includes use Cldr and therefore is a Cldr
    backend module(). The default is Money.default_backend/0.

* Any other options are considered as formatting options to be applied by
    default when calling Money.to_string/2.
  • :locale is any known locale.  The locale is used to normalize any
    binary (String) amounts to a form that can be consumed by Decimal.new/1.
    This consists of removing any localised grouping characters and replacing
    the localised decimal separator with a ".". The default is
    Cldr.get_locale/0.

**Comment by kipcole9** [comment]
> * It's true, just that passing "100.25" results in "10025".

Thats surprising since in my testing:
```elixir
iex> Money.new(:COP, "100.25") |> Jason.encode!() |> Jason.decode!() |> Money.Ecto.Composite.Type.cast()
{:ok, Money.new(:COP, "100.25")}
```

**Comment by jfpedroza** [comment]
> Thats surprising since in my testing:
>
> ```elixir
> iex> Money.new(:COP, "100.25") |> Jason.encode!() |> Jason.decode!() |> Money.Ecto.Composite.Type.cast()
> {:ok, Money.new(:COP, "100.25")}
> ```

Perhaps because the default locale is not `es-CO`?

This is what I get:

```elixir
iex> Money.new(:COP, "100.25")
Money.new(:COP, "10025")
```

**Comment by kipcole9** [comment]
Bingo. Definitely a bug:
```elixir
iex> Cldr.put_default_locale "es-CO"
{:ok, #Cldr.LanguageTag<es-CO [validated]>}

iex> Money.new(:COP, "100.25")
Money.new(:COP, "100.25")

iex> Money.new :COP, "100,25"
Money.new(:COP, "10025")

iex> Cldr.get_locale
#Cldr.LanguageTag<es-CO [validated]>

# Put specifing to locale works so its related to
# the default locale handling which should be quick
# to fix.
iex> Money.new :COP, "100,25", locale: "es-CO"
Money.new(:COP, "100.25")
```
Will investigate ASAP today, sorry for the inconvenience.

**Comment by jfpedroza** [comment]
Thanks. That output seems odd, though.

In my case it parses "100,25" properly but not "100.25". As I mentioned, COP uses `,` for decimals.

Add passing `locale: "es-CO"` directly doesn't solve the issue.

**Comment by kipcole9** [comment]
It is parsing "correctly" but not as you expect.  When parsing the string, each locale has a symbol that represents the decimal separator, and a symbol that represents the grouping separator. For `es-CO` those symbols are:
```
iex> Cldr.Number.Symbol.number_symbols_for("es-CO", Money.Cldr)
{:ok,
 %{
   latn: %Cldr.Number.Symbol{
     decimal: ",",
     group: ".",
     exponential: "E",
     infinity: "∞",
     list: ";",
     minus_sign: "-",
     nan: "NaN",
     per_mille: "‰",
     percent_sign: "%",
     plus_sign: "+",
     superscripting_exponent: "×",
     time_separator: ":"
   }
 }}
```
When parsing, the localised decimal separator is replace with the Elixir-standard decimal separator. And the **grouping separator is removed**. Since the grouping separator for `es-CO` is `.`, that is removed.  And its why you see the result you see, and not an error.

The same happens here:
```
iex> Money.new(:COP, "1.000.000,25", locale: "es-CO")
Money.new(:COP, "1000000.25")
```
Which I hope you agree is the correct result?

**Comment by jfpedroza** [comment]
Yes, that is indeed the correct result, and I already knew the `.` was simply ignored. The problem is that a `.` in that location is most likely a bug and leads to an amount that is a hundred times bigger than expected.

I would expect Money to be "smarter" and accept `.` every 3 digits before the decimal and fail otherwise. Although that is probably not easy to do, specially all the other locales.

**Comment by kipcole9** [comment]
There are a myriad of possibly problems in free form text that make deterministic parsing quite difficult. There are lots of numbers that are ill-formed in their string form and I do quite a lot of work to "make sense" of the numbers. But transforming arbitrary textual input into an amount is outside the scope of `:money`.

One possibility might be to have a "Money.validate_money_string/2` function that has a more structured approach to parsing strings that are expected to be conformant in decimal separator, grouping separator, grouping size (complicated in some locales!) and currency digits.  

I'll have a think about adding that if you think its useful, but I can't prioritise it for the next few weeks.

**Comment by jfpedroza** [comment]
I was looking for a way to prevent the bug where we receive the map from JSON and pass the amount and currency directly to `Money.new`, forgetting about the fact that it will contain a `.` instead of a `,`. It has happened a few times already.

The solution for that seems to be remembering to call `Decimal.new/1` on the amount before, so I'm not sure the function will be useful. And for the case of passing to a changeset, I'm just iterating through the data and changing any map that looks like `%{"currency" => currency, "amount" => amount}` to `%{"currency" => currency, "amount" => Decimal.new(amount)}` before calling `Changeset.cast`.

**Comment by LostKobrakai** [comment]
I think the issue here is that `Money.new` is one API, when the use-case for creating a new money struct includes both unlocalized (machine level) cases as well as cases where text provide by humans is supposed to be handled.

I know this would be a breaking change, but I would expect `Money.new` to not be locale sensitive by default and either require explicitly passing the locale to be sensitive to it or have explicit/separate API for that. I think being locale sensitive by default is great for anything outputting things, but for input there's no good way to know if things are already "normalized" or "human input".

**Comment by kipcole9** [comment]
It's a very fair point @LostKobrakai, and maybe I would do dofferently if I was starting again.  In my mind I see the following contract for Money.new/3:

1. If the amount is an integer or decimal, all ok

2. If it's a float, deprecate and then process.  Remove in the next major version.

3. If it's a strong, do the best possible to interpret the string in a localised manner.

I appreciate that (3) is quite non-deterministic. But there are now behaviours relied about for several years. Changing it would be unreasonable.

I can do the following:

1. Make the documentation much better.  Will do. At least lib users need to know what happens.

2. Provide a more strict parser (can do, non trivial, will take some time and not a particular priority)

3. Fix the bug where the global locale is not respected (will do)

There is a difficult balance here between what the scope of Money is versus what is good to do. I'm open to ideas but I'm not very motivated to move far from the current scope.  I'll consider validate_money_amount/1 or similar but after that I think more NL parsing is outside of scope.

**Comment by kipcole9** [comment]
Update on the issue regarding the default locale for parsing:

1. `Money.new(:COP, "1234,45")` will ultimately call `Money.default_backend().get_locale()` to determine the locale to be used for parsing.
2. Therefore, to set the default locale for parsing:
  * DO `Cldr.put_locale "en-CO", MyCldr.Backend`
  * DO NOT DO `Cldr.put_locale "en-CO"

I will investigate the locale fallback resolution in `:ex_cldr` because I expected that if `MyCldr.Backend` has no default locale then fallback would be to the system default locale (ie `Cldr.get_locale/0`) but it doesn't do that.  That part of `:ex_cldr` goes back to the very early days and the entire `ex_cldr_*` set of libraries depends on it so I need to go carefully if any behaviour has to change.

Sorry it took so long to work out what the parsing was not proceeding as expected on your original issue. Let me know if following the note above at least gets you the expected result?

**Comment by kipcole9** [comment]
I'll close *this* issue for now but no issues if you need to reopen.  I'll work on the "get locale" issue in `:ex_cldr` over the weekend.

**Comment by jfpedroza** [comment]
Sorry for the delay. The default locale is set at the application level in `use Cldr, default_locale: "es-CO"` and is not changed anywhere, so no issues there.

Perhaps a new function that is the equivalent of `Money.new(currency, Decimal.new(amount))` would be good? And then
point in `Money.new` that if the amount string is "normalized" or machine input, then to call that new function instead because `new` will return unexpected output.

If not, at the very least, I agree that the documentation should be improved and can consider the issue closed.

**Comment by LostKobrakai** [comment]
I just ran into this as well updating ex_cldr on a project, where the Ecto.Type yielded different results based on the user selected locale. I really think there should be an alternative to `Money.new`, which is not affected by the current locale.

**Comment by LostKobrakai** [comment]
I also noticed that this update has been from 5.2 to 5.12 / _sql 1.3 to 1.7, so I feels like a breaking change already
